content/posts/networkx/vf2pp/graph-iso-vf2pp/index.md

```diff
@@ -0,0 +1,132 @@
+3. Take the next unexamined node $u$ from the ordering.
+4. Find its candidates and check if there's a candidate $v$ such that the pair $u-v$ satisfies the **feasibility rules**
+5. if there's any, extend the mapping and **go to 3**.
+6. if not, pop the last pair $\hat{u}-\hat{v}$ from the mapping and try a different candidate $\hat{v}$, from the remaining candidates of $\hat{u}$
+7. The two graphs are **isomorphic** if the number of **mapped nodes** equals the number of nodes of the two graphs.
+8. The two graphs are **not isomorphic** if there are no remaining candidates for the first node of the ordering (root).
+
+The pseudocode for the **VF2++** is presented below.
+
+> ---
+>
+> **VF2++**
+>
+> ---
+>
+> 1. **Set** Check if the number of nodes and degree sequence is the same for both graphs.
+> 2. **Set** $T_1, T_2, \tilde{T}_1, \tilde{T}_2$ to default
+> 3. **Set** $m = m^{-1} = \{\}$
+> 4. $order=\def\textsc#1{\dosc#1\csod}{matchingOrder()}$
+> 5. $root=order\[0\]$
+> 6. **Set** $stack=\[\(root, \def\textsc#1{\dosc#1\csod}{findCandidates(root)}\)\]$
+> 7. **while** $stack$ **do**
+>    - $n,nc=stack.peek()$
+>    - **if** $nc$ is empty **do**
+>      - $stack.pop()$
+>      - $\def\textsc#1{\dosc#1\csod}{restoreState()}$
+>      - $continue$
+>    - **if** $\def\textsc#1{\dosc#1\csod}{feasibility(n, nc.pop())}$ **do**
+>      - **if** len(m) equals len(G1) **return True**
+>      - $\def\textsc#1{\dosc#1\csod}{updateState()}$
+>      - $nextNode=order.next()$
+>      - $stack.append\(\(nextNode, \def\textsc#1{\dosc#1\csod}{findCandidates(nextNode)}\)\)$
+
+## Performance
+
+This section is dedicated to the performance comparison between **VF2** and **VF2++**. The comparison was performed in
+**random graphs** without labels, for number of nodes anywhere between the range $\(100-2000\)$. The results are depicted
+in the two following diagrams.
+
+<center><img src="times.png" alt="vf2++ and vf2 times"/></center>
+<center><img src="speedup.png" alt="speedup"/></center>
+
+We notice that the maximum speedup achieved is **14x**, and it keeps increasing gradually, as the number of nodes increase.
+It is also highly prominent that the increase in number of nodes, doesn't seem to affect the performance of **VF2++** to
+a significant extent, when compared to the drastic impact on the performance of **VF2**. Our results are almost identical
+to those presented in the original **VF2++ paper**, verifying the theoretical analysis and premises of the literature.
+
+## Optimizations
+
+The achieved boost is due to some key improvements and optimizations, and more specifically:
+
+- **Optimal node ordering**, which avoids following unfruitful branches that will result in infeasible states. We make sure that the nodes that have the biggest possibility to match, are accessed first.
+- **Implementation in a non-recursive manner**, which saves both time and space.
```

Comment:
Recursive implementations are not *always* generally slower than iterative ones, so I'd vote to be a little more specific here. IMO the main advantages of an iterative implementation for this case are:
 - Avoiding Python's maximum recursion depth limit
 - Reducing function call overhead